### DISCUSSION TITLE
How to replace Basic Authentication of usrid & pwd with the JWT token

### DISCUSSION THREAD

**Discussion by ghost** [zowe/zowe-cli#2183] [Q&A] [answered]
I successfully configured this npm with basic authentication, with a user id and password. Now i would like to replace with JWT tokens from z/osmf.  Is there is a sample syntax program for it pls?

**Comment by traeok** [comment, marked as answer]
Hi @inv9wj1,

For clarification, are you developing a program using Zowe CLI (or its SDKs)? Or are you using Zowe CLI with a Team Configuration file set up for your mainframe instance?

If you are developing using Zowe CLI/SDKs, you can provide the JWT token in the `tokenValue` property of a session, and then you can use that session with our available APIs. For example, with the z/OS Files SDK:
```ts
const profInfo = new ProfileInfo("zowe");
await profInfo.readProfilesFromDisk();
const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, { getSecureVals: true });
const session = imperative.ProfileInfo.createSession(zosmfMergedArgs.knownArgs);

// Specify JWT token type and value on the session object.
session.ISession.tokenType = "jwtToken";
session.ISession.tokenValue = "<your JWT here>";

const dataset = "ZOWEUSER.PUBLIC.YOUR.DATASET.HERE";
const options: IDownloadOptions = { failFast: false };
const response = await Download.allMembers(session, dataset, options);
```

If you are using a z/OSMF profile with Zowe CLI and want to authenticate using JWT, you can specify a token type and token value on your profile. The provided token will be used to authenticate future requests.

**Note:** If your environment supports using secure credentials with Zowe CLI, I would recommend saving the properties securely so that they are not shown in plaintext. Use the following Zowe CLI commands to store the `tokenType` and `tokenValue` securely:

```
zowe config set "profiles.zosmf.properties.tokenType" "jwtToken"
zowe config set "profiles.zosmf.properties.tokenValue" "<your JWT here>" --secure
```

Otherwise, you can provide the `tokenType` and `tokenValue` properties on the profile, directly within the Team Configuration file:

```json
"zosmf": {
    "type": "zosmf",
    "properties": {
        "tokenType": "jwtToken",
        "tokenValue": "<your JWT here>"
    }
}
```

**Reply by ghost** [reply]
Thanks Much for the response. 

I am using zowe/CLI SDK's, and i think, the below would suffice my requirement. 
Again, Appreciate the prompt response, it really helps us. 

const profInfo = new ProfileInfo("zowe");
await profInfo.readProfilesFromDisk();
const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, { getSecureVals: true });
const session = imperative.ProfileInfo.createSession(zosmfMergedArgs.knownArgs);

// Specify JWT token type and value on the session object.
session.ISession.tokenType = "jwtToken";
session.ISession.tokenValue = "<your JWT here>";

const dataset = "ZOWEUSER.PUBLIC.YOUR.DATASET.HERE";
const options: IDownloadOptions = { failFast: false };
const response = await Download.allMembers(session, dataset, options);

**Reply by traeok** [reply]
No problem! Yes, this example code should be sufficient for authenticating over JWT. Feel free to follow up here if you have any questions or concerns!

**Reply by ghost** [reply]
Hello Again, 
IBM asked me to use Ltpa Token from the cookie, as there seems to be an issue with jwt generation in my account mainframe. Where i used the below syntax and am getting into error while creating session. 

It looks like it did returns a ltpa token, but unable to create a session with that, could you help me on this pls:

**Error msg** :
TypeError: Cannot read properties of null (reading 'profLoc')
 at ProfileInfo.mergeArgsForProfile (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\config\src\ProfileInfo.js:540:21)
    at createSession (C:\Users\F8512PV\Desktop\gitlab\zoops_server\routes\mainframe.js:412:36)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async C:\Users\F8512PV\Desktop\gitlab\zoops_server\routes\mainframe.js:434:21
GET /mf/download-dataset?dataset=sysshr.zoops.verbose 500 1470.203 ms - 25



**Response headers from zosmf :**

Response headers: Object [AxiosHeaders] {
  'x-powered-by': 'Servlet/3.1',
  'x-content-type-options': 'nosniff',
  'x-xss-protection': '1; mode=block',
  'content-type': 'application/json;charset=ISO-8859-1',
  'content-language': 'en-US',
  'strict-transport-security': 'max-age=31536000; includeSubDomains',
  'set-cookie': [
    'LtpaToken2=dTT/8H+w1WlZcwcw+Ibp6Txq1lfDBIdhdZgfh2MEbxAtOoOc/7SiOqhOL7bJrIDT77cP8/Rb6ay1VTTjGERfrhCdy5foynP1pXIo/67EtAcW1qx9uXE7Ne6zcmKWWO1vCbHL3R+zsjo91lgj1gjef/T0fLxheEumfkLnAOWwIHn5S5oq+DT8AnOvo0r3cX8ZG8kdtJs4l820YVPSIw4Dj6+M9TtRdm/mxGTsj8NW6M4apdylLE+mQYN3/uU4DjXW5/9JwJiiSsC0zxb9y1JLPqP8xHMrYKpdA/ZbIjZrYNRDJY+tvPY+eBRNK1m4piTe; Path=/; Secure; HttpOnly'
  ],
  'transfer-encoding': 'chunked',
  connection: 'Close',
  date: 'Mon, 01 Jul 2024 11:20:50 GMT',
  expires: 'Thu, 01 Dec 1994 16:00:00 GMT',
  'cache-control': 'no-cache="set-cookie, set-cookie2"'
}

**parsed ltpa token from prgram**
dTT/8H+w1WlZcwcw+Ibp6Txq1lfDBIdhdZgfh2MEbxAtOoOc/7SiOqhOL7bJrIDT77cP8/Rb6ay1VTTjGERfrhCdy5foynP1pXIo/67EtAcW1qx9uXE7Ne6zcmKWWO1vCbHL3R+zsjo91lgj1gjef/T0fLxheEumfkLnAOWwIHn5S5oq+DT8AnOvo0r3cX8ZG8kdtJs4l820YVPSIw4Dj6+M9TtRdm/mxGTsj8NW6M4apdylLE+mQYN3/uU4DjXW5/9JwJiiSsC0zxb9y1JLPqP8xHMrYKpdA/ZbIjZrYNRDJY+tvPY+eBRNK1m4piTe

**code syntax** 

`const httpsAgent = new https.Agent({
  rejectUnauthorized: false,
});

let ltpaToken = null;
let tokenExpirationTime = null;

async function getLtpaToken() {
  const auth = Buffer.from(`${user}:${password}`).toString("base64");

  try {
    const response = await axios.post(
      `https://${zosmfHost}:${zosmfPort}/zosmf/services/authenticate`,
      null,
      {
        headers: {
          Authorization: `Basic ${auth}`,
          "X-CSRF-ZOSMF-HEADER": "HeaderValue", // Add your actual header value here
        },
        httpsAgent,
        withCredentials: true, // Ensure that axios includes cookies in the request
      }
    );

    // Log the response headers to see the cookies
    console.log("Response headers:", response.headers);

    const setCookieHeader = response.headers["set-cookie"];

    if (!setCookieHeader) {
      throw new Error("Set-Cookie header not found in response");
    }

    // Find the LTPA token in the cookies
    const ltpaCookie = setCookieHeader.find((cookie) =>
      cookie.startsWith("LtpaToken2=")
    );
    console.log("ltpaToken2 is", ltpaCookie);
    if (!ltpaCookie) {
      throw new Error("LTPA token not found in cookies");
    }

    // Extract the token value from the cookie
    ltpaToken = ltpaCookie.split(";")[0].split("=")[1];
    console.log("ltpaToken is", ltpaToken);
    // Optionally, you can parse and handle the expiration time if available
    // Example: assuming the cookie contains an expiration time
    // const expirationMatch = ltpaCookie.match(/Expires=(.*?);/);
    // if (expirationMatch) {
    //     tokenExpirationTime = new Date(expirationMatch[1]).getTime();
    // }

    return ltpaToken;
  } catch (error) {
    console.error("Error obtaining LTPA token:", error);
    throw error;
  }
}

async function createSession() {
  const currentTime = Date.now();

  if (
    !ltpaToken ||
    (tokenExpirationTime && currentTime >= tokenExpirationTime)
  ) {
    console.log("LTPA token expired or not found, obtaining a new token.");
    await getLtpaToken();
  }

  const profInfo = new ProfileInfo("zowe");
  await profInfo.readProfilesFromDisk();
  const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
  const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, {
    getSecureVals: true,
  });
  const session = imperative.ProfileInfo.createSession(
    zosmfMergedArgs.knownArgs
  );

  // Specify LTPA token type and value on the session object.
  session.ISession.tokenType = "ltpaToken";
  session.ISession.tokenValue = ltpaToken;

  return session;
}

router.get("/download-dataset", async (req, res) => {
  const datasetName = req.query.dataset; // get dataset name from query parameters

  if (!datasetName) {
    return res.status(400).send("Dataset name is required");
  }

  try {
    const session = await createSession();
    const options = { failFast: false };
    const response = await Download.allMembers(session, datasetName, options);
    res.setHeader(
      "Content-Disposition",
      `attachment; filename="${datasetName}.txt"`
    );
    res.setHeader("Content-Type", "text/plain");
    res.send(response.toString());
  } catch (error) {
    console.error(error);
    res.status(500).send("Error downloading dataset");
  }
});
`

**Reply by traeok** [reply]
Hi @inv9wj1,

I noticed that the `tokenType` in the code above is set to `ltpaToken` instead of `LtpaToken2`. I would recommend specifying the token type as `LtpaToken2`, as we have specific checks in place to handle certain token types (such as the LTPA2 token format). You can also refer to the variables defined in [`SessConstants.ts`](https://github.com/zowe/zowe-cli/blob/master/packages/imperative/src/rest/src/session/SessConstants.ts#L77) to access the supported token types. 

Based on the error, it seems that `zosmfProfAttrs` is `null` when passed into `profInfo.mergeArgsForProfile`.
Note that the `getDefaultProfile` function may return `null` instead of a `IProfAttrs` object when it cannot find a default profile for the given type. The `mergeArgsForProfile` function expects a valid `IProfAttrs` object and will fail if provided a falsy value (such as `null`).

Can you confirm whether you have a default profile for `zosmf` defined in your Team Config?

Example:
```jsonc
{
    "profiles": {
        /* Zowe profiles are defined here */
        "zosmfprofile": {
            "type": "zosmf",
            /* connection info and properties for zosmfprofile are defined here */
        }
    },
    "defaults": {
        "zosmf": "zosmfprofile"
    },
    "autoStore": true
}
```

**Reply by ghost** [reply]
no, i am afraid, i do not have a default profile, could you share a link, explains how to create and connect it pls?...

**Reply by traeok** [reply]
Hi @inv9wj1,

Please refer to the article "Team configuration for application developers" to learn more about how to define a Team Config and specify its defaults: https://docs.zowe.org/stable/user-guide/cli-using-team-configuration-application-developers

The section "Editing team configurations" at the bottom of the page has an example of a Team Config with a default `zosmf` profile.

**Reply by ghost** [reply]
Thank you, i added a ~.zowe/zowe.config.json with below data, it seems the step passed that point, but ended up with something else with the same profileInfo.ts module.

Could you check the below error msg, and share your suggestions pls. 

content inside ~.zowe/zowe.config.json
```
{
  "profiles": {
    "zosmf": {
      "type": "zosmf",
      "properties": {
        "host": "xxxxx.1dc.com",
        "rejectUnauthorized": false,
        "port": 1443
      }
    }
  },
  "defaults": {
    "zosmf": "zosmf"
  },
  "plugins": []
}
```
error msg:
```

ProfInfoErr: Failed to initialize secure credential manager
    at ProfileInfo.<anonymous> (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\config\src\ProfileInfo.js:839:23)
    at Generator.throw (<anonymous>)
    at rejected (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\config\src\ProfileInfo.js:16:65)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  mDetails: {
    errorCode: 'LoadCredMgrFailed',
    msg: 'Failed to initialize secure credential manager',
    causeErrors: ImperativeError: Failed to load Keytar module: Cannot find module 'keytar'
    Require stack:
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\DefaultCredentialManager.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\CredentialManagerFactory.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\index.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\cmd\src\profiles\CliProfileManager.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\cmd\index.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\index.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\methods\copy\Copy.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\methods\copy\index.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\index.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\routes\mainframe.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\app.js
    - C:\Users\F8512PV\Desktop\gitlab\zoops_server\bin\www
        at DefaultCredentialManager.<anonymous> (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\DefaultCredentialManager.js:110:34)
        at Generator.next (<anonymous>)
        at C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\DefaultCredentialManager.js:18:71
        at new Promise (<anonymous>)
        at __awaiter (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\DefaultCredentialManager.js:14:12)
        at DefaultCredentialManager.initialize (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\DefaultCredentialManager.js:96:16)
        at CredentialManagerFactory.<anonymous> (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\CredentialManagerFactory.js:134:41)
        at Generator.next (<anonymous>)
        at C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\security\src\CredentialManagerFactory.js:18:71
        at new Promise (<anonymous>) {
      mDetails: [Object],
      mMessage: "Failed to load Keytar module: Cannot find module 'keytar'\n" +
        'Require stack:\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\security\\src\\DefaultCredentialManager.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\security\\src\\CredentialManagerFactory.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\security\\index.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\cmd\\src\\profiles\\CliProfileManager.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\cmd\\index.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\imperative\\lib\\index.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\zos-files-for-zowe-sdk\\lib\\methods\\copy\\Copy.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\zos-files-for-zowe-sdk\\lib\\methods\\copy\\index.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\node_modules\\@zowe\\zos-files-for-zowe-sdk\\lib\\index.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\routes\\mainframe.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\app.js\n' +
        '- C:\\Users\\F8512PV\\Desktop\\gitlab\\zoops_server\\bin\\www'
    }
  },
  mMessage: 'Failed to initialize secure credential manager',
  mItemsInError: []
}
```

**Reply by traeok** [reply]
The given Imperative error suggests that the `ProfileInfo` class was not able to initialize the secure credential manager. If you would like to use the secure credential manager, install the `@zowe/secrets-for-zowe-sdk` package as a dependency of your application. Once installed, Imperative will locate and initialize it when calling the `ProfileInfo.readProfilesFromDisk` function.

**Reply by ghost** [reply]
I tried installing the @zowe/secrets-for-zowe-sdk , and it seems forbidden in our policy, wondering why zowe has been in that list. 
Could you help me on this pls, is there any way to fix this?

npm ERR! code E403
npm ERR! 403 403 Forbidden - GET https://registry.npmjs.org/@zowe%2fsecrets-for-zowe-sdk
npm ERR! 403 In most cases, you or one of your dependencies are requesting
npm ERR! 403 a package version that is forbidden by your security policy, or
npm ERR! 403 on a server you do not have access to.

npm ERR! A complete log of this run can be found in: C:\Users\F8512PV\AppData\Local\npm-cache\_logs\2024-07-02T14_36_23_564Z-debug-0.log

**Reply by traeok** [reply]
@inv9wj1, are you using the offline bundle of packages from Zowe.org? If so, previous versions of the Zowe bundle did not include the Secrets SDK due to an oversight during packaging. The `@zowe/secrets-for-zowe-sdk` package will be available in the upcoming Zowe 2.17 bundle. My apologies for any inconvenience caused by this.

However, if you are not using the offline bundle and are having issues fetching Zowe dependencies directly from NPM, then this issue is related to networking or security restrictions at your company site. In that case, I would recommend reaching out to your security team for next steps.

Thanks!

**Reply by ghost** [reply]
Thanks for the assistance, 
i have downloaded from nexus repo present in my org and added the npm @zowe/secrets-for-zowe-sdk. And added the zowe.config.json and zowe.schema.json. It returns the below error. Could you help fix this pls?

Error msg :
```
profileInfo is [class ProfileInfo]    --> from console.log('profileInfo is',profileInfo);
TypeError: Cannot read properties of undefined (reading 'ProfileInfo')
    at createSession (C:\Users\F8512PV\Desktop\gitlab\zoops_server\routes\mainframe.js:427:30)
    at async C:\Users\F8512PV\Desktop\gitlab\zoops_server\routes\mainframe.js:446:21
GET /mf/download-dataset?dataset=sysshr.zoops.verbose 500 1425.200 ms - 25

```

Code :
```
async function createSession() {
  const currentTime = Date.now();

  if (
    !ltpaToken ||
    (tokenExpirationTime && currentTime >= tokenExpirationTime)
  ) {
    console.log("LTPA token expired or not found, obtaining a new token.");
    await getLtpaToken();
  }

  const profInfo = new ProfileInfo("zowe");
  await profInfo.readProfilesFromDisk();
  const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
  const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, {
    getSecureVals: true,
  });
  console.log("profileInfo is", ProfileInfo);
  const session = imperative.ProfileInfo.createSession(
    zosmfMergedArgs.knownArgs
  );

  // Specify LTPA token type and value on the session object.
  session.ISession.tokenType = "LtpaToken2";
  session.ISession.tokenValue = ltpaToken;

  return session;
}

```

zowe.config.json:
```

{
    "$schema": "./zowe.schema.json",
    "profiles": {
        "zosmf": {
            "type": "zosmf",
            "properties": {
                "host": "xxxxxxx.1dc.com",
                "rejectUnauthorized": false,
                "port": 443
            },
            // "secure": [
            //     "user",
            //     "password",
            // ]
        }
    },
    "defaults": {
        "zosmf": "zosmf",
    },
    "plugins": []
}

```

**Reply by traeok** [reply]
Hi @inv9wj1,

From the given error, its unclear as to which line is the culprit - but from looking at the `createSession` function, I can see that you're trying to create an instance of the `ProfileInfo` class. If you do not have the `ProfileInfo` class imported, it will fail with a similar error inside that function. I would recommend double-checking your import statements and require calls to ensure that you have access to the `ProfileInfo` class wherever `createSession` is defined. Hope this helps!

Note that if you plan on using the Secrets SDK to store and load credentials, they will be stored on the system where the application is being hosted. For example, if your project is a Node.js server application and does not run on an end-user's machine directly, the credentials will be accessed and stored on the server itself.

**Reply by ghost** [reply]
Hi @traeok , 
I am not intent to store the login credentials directly, instead i have a react application, which sends the login request to the node js server, where the node js server, just receives and authenticates and then it saves the LtpaToken2 for a particular time, then it creates session using the LtpaToken2 and gets the data out of mainframe and sends back to react application. 

with the next set of requests, the node js continues to authenticate the user using the store ltpatoken2 and process the request. 

Earlier i used basic authentication method and worked fine, while am trying to work with LtpaToken2, am getting into issues. 

Below is the actual code present in Node js for your reference, could you check , does it have the right syntax pls?... am unable to verify as there is no error related to the syntax. also didnt get any errors related to imports. 

code from node js:

```
const { ProfileInfo, imperative } = require("@zowe/imperative");
// Zowe session configuration request using JWT
const zosmfHost = "xxxxxxxxx.1dc.com"; // replace with your z/OSMF host
const zosmfPort = 443; // replace with your z/OSMF port
const userr = "xxxxxxx"; // replace with your z/OSMF username
const passwordd = "yyyyyyyyy"; // replace with your z/OSMF password


const httpsAgent = new https.Agent({
  rejectUnauthorized: false,
});

let ltpaToken = null;
let tokenExpirationTime = null;

async function getLtpaToken() {
  const auth = Buffer.from(`${userr}:${passwordd}`).toString("base64");

  try {
    const response = await axios.post(
      `https://${zosmfHost}:${zosmfPort}/zosmf/services/authenticate`,
      null,
      {
        headers: {
          Authorization: `Basic ${auth}`,
          "X-CSRF-ZOSMF-HEADER": "dummy", // Add your actual header value here
        },
        httpsAgent,
        withCredentials: true, // Ensure that axios includes cookies in the request
      }
    );

    // Log the response headers to see the cookies
    console.log("Response headers:", response.headers);

    const setCookieHeader = response.headers["set-cookie"];

    if (!setCookieHeader) {
      throw new Error("Set-Cookie header not found in response");
    }

    // Find the LTPA token in the cookies
    const ltpaCookie = setCookieHeader.find((cookie) =>
      cookie.startsWith("LtpaToken2=")
    );
    console.log("ltpaToken2 is", ltpaCookie);
    if (!ltpaCookie) {
      throw new Error("LTPA token not found in cookies");
    }

    // Extract the token value from the cookie
    ltpaToken = ltpaCookie.split(";")[0].split("=")[1];
    console.log("ltpaToken is", ltpaToken);
    // Optionally, you can parse and handle the expiration time if available
    // Example: assuming the cookie contains an expiration time
    // const expirationMatch = ltpaCookie.match(/Expires=(.*?);/);
    // if (expirationMatch) {
    //     tokenExpirationTime = new Date(expirationMatch[1]).getTime();
    // }

    return ltpaToken;
  } catch (error) {
    console.error("Error obtaining LTPA token:", error);
    throw error;
  }
}

async function createSession() {
  const currentTime = Date.now();

  if (
    !ltpaToken ||
    (tokenExpirationTime && currentTime >= tokenExpirationTime)
  ) {
    console.log("LTPA token expired or not found, obtaining a new token.");
    await getLtpaToken();
  }

  const profInfo = new ProfileInfo("zowe");
  await profInfo.readProfilesFromDisk();
  const zosmfProfAttrs = profInfo.getDefaultProfile("zosmf");
  const zosmfMergedArgs = profInfo.mergeArgsForProfile(zosmfProfAttrs, {
    getSecureVals: true,
  });
  console.log("profileInfo is", ProfileInfo);
  const session = imperative.ProfileInfo.createSession(
    zosmfMergedArgs.knownArgs
  );

  // Specify LTPA token type and value on the session object.
  session.ISession.tokenType = "LtpaToken2";
  session.ISession.tokenValue = ltpaToken;

  return session;
}

router.get("/download-dataset", async (req, res) => {
  const datasetName = req.query.dataset; // get dataset name from query parameters

  if (!datasetName) {
    return res.status(400).send("Dataset name is required");
  }

  try {
    const session = await createSession();
    const options = { failFast: false };
    const response = await Download.allMembers(session, datasetName, options);
    res.setHeader(
      "Content-Disposition",
      `attachment; filename="${datasetName}.txt"`
    );
    res.setHeader("Content-Type", "text/plain");
    res.send(response.toString());
  } catch (error) {
    console.error(error);
    res.status(500).send("Error downloading dataset");
  }
});

```

**Reply by traeok** [reply]
@inv9wj1,

I noticed that the first line is trying to grab the `ProfileInfo` export from the Imperative package, and it also tries to grab the "imperative" export. While this will work to grab the ProfileInfo class, the "imperative" named export does not exist in the Imperative package, so it will be undefined:

![image](https://github.com/zowe/zowe-cli/assets/86500639/6fada76b-b8eb-465d-97e4-98f74cdcd2eb)

This is why the error occurs - in the `createSession` function, `session` is created by calling `imperative.ProfileInfo.createSession`, and as imperative is undefined, `ProfileInfo` does not exist as a property. You can fix this by changing that line to `ProfileInfo.createSession`, or by updating the import:

![image](https://github.com/zowe/zowe-cli/assets/86500639/68829d91-b8d6-480d-9d7f-db4c9ba4911f)

---

To clarify, our examples are intended for desktop apps and not web apps. The "ProfileInfo" facilities use I/O to read and write profile changes to your Team Config file, and its primary use is for plug-ins and extenders that directly run on the user's machine, rather than a server environment. 

If you do not want to store the credentials securely and do not intend to save this info in a Zowe config, you can modify a session object that's already been configured to use basic authentication. First, you must set `session.ISession.tokenType` to `LtpaToken2` and set `session.ISession.tokenValue` to the LTPA token. Then, you can delete the `session.ISession.user` and `session.ISession.password` properties, if they exist, so that it will use the token rather than basic authentication to perform the request.

**Reply by ghost** [reply]
Hi @traeok,
Thanks for the detailed explanation, Based on your answer, looks like the code i had is for desktop apps.  
My issue is, i have a react app, and node js, and in node Js, i have hard coded the user id and password in a .env file and i am now working to upgrade to have a proper authentication method.

Just one more question, Could you share the full syntax to authenticate using LtpaToken2 for web applications pls?

**Reply by traeok** [reply]
@inv9wj1:
For a web application, I would recommend skipping profiles altogether and creating a `Session` object (defined in Imperative) to pass to the SDKs.

To construct a Session object, please refer to the typedoc for Imperative:
- `Session` class: https://docs.zowe.org/stable/typedoc/classes/_zowe_imperative.session#constructor
- `ISession` interface (defines supported properties for `Session` constructor): https://docs.zowe.org/stable/typedoc/interfaces/_zowe_imperative.isession

Here's an example of how you would authenticate using the constructed Session object. I've updated your code snippet for `createSession` and provided a brief example of using it with the `Download.allMembers` function.

```ts
import { Session } from "@zowe/imperative";

async function createSession() {
  const currentTime = Date.now();

  if (
    !ltpaToken ||
    (tokenExpirationTime && currentTime >= tokenExpirationTime)
  ) {
    console.log("LTPA token expired or not found, obtaining a new token.");
    await getLtpaToken();
  }

  return new Session({
    hostname: "<YourHostHere>",
    port: 1234,
    // Specify LTPA token type and value on the session object.
    tokenType: "LtpaToken2",
    tokenValue: ltpaToken
  });
}

// Making the request:
const session = await createSession();
const options = { failFast: false };
const response = await Download.allMembers(session, "YOUR.PDS", options);
```

**Note:** Depending on the LPAR you're trying to connect to, you might need to pass more options to the session such as `rejectUnauthorized` in order to establish a successful connection. Refer to the `ISession` interface typedoc linked above to see all available properties.

**Reply by ghost** [reply]
@traeok
I tried the syntax and ends with error, saying looking for credentials parm. i am checking the docs, meanwhile, anyidea why its looking for credentials pls?, while we have ltpatoken.

code :

```
async function createSession() {
  const currentTime = Date.now();

  if (
    !ltpaToken ||
    (tokenExpirationTime && currentTime >= tokenExpirationTime)
  ) {
    console.log("LTPA token expired or not found, obtaining a new token.");
    await getLtpaToken();
  }

  return new Session({
    hostname: "tplavipa.1dc.com",
    port: 443,
    // Specify LTPA token type and value on the session object.
    tokenType: "LtpaToken2",
    tokenValue: ltpaToken,
  });
}
router.get("/download-dataset", async (req, res) => {
  const datasetName = req.query.dataset; // get dataset name from query parameters

  if (!datasetName) {
    return res.status(400).send("Dataset name is required");
  }

  try {
    const session = await createSession();
    const options = { failFast: false };
    const response = await Download.allMembers(session, datasetName, options);
    res.setHeader(
      "Content-Disposition",
      `attachment; filename="${datasetName}.txt"`
    );
    res.setHeader("Content-Type", "text/plain");
    res.send(response.toString());
  } catch (error) {
    console.error(error);
    res.status(500).send("Error downloading dataset");
  }
});

```


Error msg:
```

ImperativeError: No credentials for a BASIC or TOKEN type of session.
    at ZosmfRestClient.buildOptions (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:372:19)
    at C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:132:39
    at new Promise (<anonymous>)
    at ZosmfRestClient.request (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\rest\src\client\AbstractRestClient.js:115:16)
    at ZosmfRestClient.getExpectString (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\imperative\lib\rest\src\client\RestClient.js:196:34)
    at List.<anonymous> (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\methods\list\List.js:72:72)
    at Generator.next (<anonymous>)
    at C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\methods\list\List.js:18:71
    at new Promise (<anonymous>)
    at __awaiter (C:\Users\F8512PV\Desktop\gitlab\zoops_server\node_modules\@zowe\zos-files-for-zowe-sdk\lib\methods\list\List.js:14:12) {
  mDetails: { msg: 'No credentials for a BASIC or TOKEN type of session.' },
  mMessage: 'No credentials for a BASIC or TOKEN type of session.'

```

**Reply by t1m0thyj** [reply]
@inv9wj1 Hello, when creating the session object, please try adding `type: "token"` as in the example below:
```
  return new Session({
    hostname: "tplavipa.1dc.com",
    port: 443,
    // Specify LTPA token type and value on the session object.
    type: "token",
    tokenType: "LtpaToken2",
    tokenValue: ltpaToken,
  });
```

The session needs to have a type specified: "basic", "token", "cert-pem" etc in order to determine whether basic or token auth should be used.

**Reply by ghost** [reply]
@traeok  Thank you, i got the below error, Where this occured earlier as well, and fixed it by creating the profile .zowe, but later we understood thats for desktop applications. How would we fix this for web applications pls?

```
'z/OSMF REST API Error:\n' +
      'http(s) request error event called\n' +
      'Error: unable to verify the first certificate',
    causeErrors: Error: unable to verify the first certificate
        at TLSSocket.onConnectSecure (node:_tls_wrap:1600:34)
        at TLSSocket.emit (node:events:517:28)
        at TLSSocket._finishInit (node:_tls_wrap:1017:8)
        at ssl.onhandshakedone (node:_tls_wrap:803:12) {
      code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'

```

**Reply by t1m0thyj** [reply]
@inv9wj1 Hello, if your mainframe uses self-signed certificates, please try adding `rejectUnauthorized: false` when creating the session object as in the example below:
```
  return new Session({
    hostname: "tplavipa.1dc.com",
    port: 443,
    rejectUnauthorized: false,
    // Specify LTPA token type and value on the session object.
    type: "token",
    tokenType: "LtpaToken2",
    tokenValue: ltpaToken,
  });
```

**Reply by ghost** [reply]
Thank you So much @traeok  :) it worked 👍 .. 

I heartfully appreciate the patience in working with me and fixing this.

**Reply by traeok** [reply]
No problem @inv9wj1, we're happy to help!
